crates/lib/src/transaction/validator.rs

```diff
@@ -239,63 +247,149 @@ impl TransactionValidator {
 
-            if source_account.owner != token_program.program_id() {
-                continue;
-            }
+                let token_state =
+                    token_program.unpack_token_account(&source_account.data).map_err(|e| {
+                        KoraError::InvalidTransaction(format!("Invalid token account: {}", e))
+                    })?;
 
-            if token_state.amount() < amount {
-                continue;
-            }
+                if source_account.owner != token_program.program_id() {
+                    continue;
+                }
 
-            if !validation.allowed_spl_paid_tokens.contains(&token_state.mint().to_string()) {
-                continue;
-            }
+                // Check Token2022 specific restrictions
+                if let Some(token2022_account) =
+                    token_state.as_any().downcast_ref::<Token2022Account>()
+                {
+                    // Check if token is non-transferable
+                    if token2022_account.non_transferable.is_some() {
+                        return Err(KoraError::InvalidTransaction(
+                            "Cannot transfer non-transferable token".to_string(),
+                        ));
+                    }
+
+                    // Check CPI guard
+                    if let Some(cpi_guard) = &token2022_account.cpi_guard {
+                        if cpi_guard.lock_cpi.into() {
+                            return Err(KoraError::InvalidTransaction(
+                                "Token transfer blocked by CPI guard".to_string(),
+                            ));
+                        }
+                    }
+
+                    // Check confidential transfers
+                    if token2022_account.confidential_transfer.is_some() {
+                        return Err(KoraError::InvalidTransaction(
+                            "Confidential transfers not supported".to_string(),
+                        ));
+                    }
+
+                    // Calculate actual transfer amount considering all extensions
+                    let mut actual_amount = amount;
+
+                    // Apply transfer fee if present
+                    if let Some(fee_config) = &token2022_account.transfer_fee {
+                        let basis_points = u16::from_le_bytes(
+                            fee_config.newer_transfer_fee.transfer_fee_basis_points.0,
+                        );
+                        let fee = std::cmp::min(
+                            (amount as u128 * basis_points as u128 / 10000) as u64,
+                            u64::from(fee_config.newer_transfer_fee.maximum_fee),
+                        );
+                        actual_amount = actual_amount.saturating_sub(fee);
+                    }
+
+                    // Apply interest if present
+                    if let Some(interest_config) = &token2022_account.interest_bearing {
```

Comment:
I think its fair to assume the amount requested will be in raw amounts and not the amount including interest. calculate_token_value_in_lamports should use a price source that is based on the raw amount and not based on the compounding interest